op-chain-ops/deployer/pipeline/opchain.go

```diff
@@ -91,9 +95,75 @@ func DeployOPChain(ctx context.Context, env *Env, bundle ArtifactsBundle, intent
 		DelayedWETHPermissionlessGameProxyAddress: dco.DelayedWETHPermissionlessGameProxy,
 	})
 
+	currentBlock, _ := env.L1Client.BlockNumber(ctx)
+	block, _ := env.L1Client.BlockByNumber(ctx, big.NewInt(int64(currentBlock)))
+	currentBlockHash := block.Hash()
+
+	// If any of the implementations addresses (excluding OpcmProxy) are empty,
+	// we need to set them using the implementation address read from their corresponding proxy.
+	// The reason these might be empty is because we're only invoking DeployOPChain.s.sol as part of the pipeline.
+	// TODO: Need to initialize 'mipsSingletonAddress' and 'preimageOracleSingletonAddress'
+	setImplementationAddressTasks := []func(){
+		func() {
+			setEIP1967ImplementationAddress(ctx, env.L1Client, dco.DelayedWETHPermissionedGameProxy, currentBlockHash, &st.ImplementationsDeployment.DelayedWETHImplAddress)
+		},
+		func() {
+			setEIP1967ImplementationAddress(ctx, env.L1Client, dco.OptimismPortalProxy, currentBlockHash, &st.ImplementationsDeployment.OptimismPortalImplAddress)
+		},
+		func() {
+			setEIP1967ImplementationAddress(ctx, env.L1Client, dco.SystemConfigProxy, currentBlockHash, &st.ImplementationsDeployment.SystemConfigImplAddress)
+		},
+		func() {
+			setRDPImplementationAddress(ctx, env.L1Client, dco.AddressManager, &st.ImplementationsDeployment.L1CrossDomainMessengerImplAddress)
+		},
+		func() {
+			setEIP1967ImplementationAddress(ctx, env.L1Client, dco.L1ERC721BridgeProxy, currentBlockHash, &st.ImplementationsDeployment.L1ERC721BridgeImplAddress)
+		},
+		func() {
+			setEIP1967ImplementationAddress(ctx, env.L1Client, dco.L1StandardBridgeProxy, currentBlockHash, &st.ImplementationsDeployment.L1StandardBridgeImplAddress)
+		},
+		func() {
+			setEIP1967ImplementationAddress(ctx, env.L1Client, dco.OptimismMintableERC20FactoryProxy, currentBlockHash, &st.ImplementationsDeployment.OptimismMintableERC20FactoryImplAddress)
+		},
+		func() {
+			setEIP1967ImplementationAddress(ctx, env.L1Client, dco.DisputeGameFactoryProxy, currentBlockHash, &st.ImplementationsDeployment.DisputeGameFactoryImplAddress)
+		},
+	}
```

Comment:
I think there's a better approach you can use here that leverages channels.

1. Create a buffered channel `make(chan error, len(tasks))`
2. Iterate over your tasks, and pass the channel into each of the inner task methods (e.g., `setEIP1967...()`). Call `go func...` directly from the for loop, there's no need for a separate `executeTasks` method.
3. Inside the task method, send either the error or `nil` to the channel.
4. Using a for loop, collect the error responses.
5. Set any errors to an error variable on the first error you see. Make sure the for loop continues until all responses have been collected.
6. Return either the error, or nil.

Happy to pair on this if you'd like an example.